### ISSUE TITLE
reindex can produce very large failures in API response

### ISSUE THREAD

**Issue by djschny** [elastic/elasticsearch#20199]
**Elasticsearch version**: 5.0.0-alpha5

**Plugins installed**: []

**JVM version**: 1.8.0_92-b14

**OS version**: Mac 10.11.5

**Description of the problem including expected versus actual behavior**:
The `_reindex` and `_update_by_query` APIs return errors in the API response by default if some are found. Unfortunately this can be extremely large, for example sometimes when first experimenting and running large processor pipelines over large production data as part of reindexing it can be common more most (greater than 50%) of the documents to actually fail. This makes for a massive API response that is not good for Elasticsearch to hold in memory itself or for the client to handle.

Some options/thoughts that come to mind are:
* default to `wait_for_completion=false` so by default huge payloads are not returned to the user
* when recording the failures, instead of embedding them in an array inside the task, instead have them as separate documents in the index. This would allow an end user to efficiently use a scroll to rip through the errors and process them in an efficient manner. Would also enable finer grained querying of the saved task errors.

**Comment by nik9000** [comment]
Maybe we should chop it to the first 50 failures or so? Something you can control with a URL parameter?

**Comment by djschny** [comment]
> Maybe we should chop it to the first 50 failures or so?

My fear with that is with the semantics right now when `?wait-for-completion=true` is that all the rest would get lost. Unless you're proposing that it automatically flips to storing in `.tasks` if payload is over 50?

Actually the more I think about this, the more I think it might be safest to always store the result in the `.tasks` index regardless of the `wait_for_completion` setting. Thoughts?

**Comment by clintongormley** [comment]
I'd be OK with limiting it to the first 50 failures (`... and $x more`).  I don't think we should return anything different in the API from what is stored in `.tasks`.

@djschny for testing purposes, just reduce the `size` parameter.

**Comment by nik9000** [comment]
Yeah - my proposal is to chop it to the first 50 failures in the response. We'd log all failures but only return the first 50.

I agree that that `.tasks` index should exactly match the result we return.

**Comment by djschny** [comment]
Chatted with @nik9000 and it looks like chopping at 50 will suffice and those wanting to look at all of the failures can centrally get that information from the logs of the coordinating node handling the API call. More graceful handling of all of this can be added in the future as usage patterns become more known.

**Comment by Mpdreamz** [comment]
this ties into the discussion here: https://github.com/elastic/elasticsearch/issues/17545

But the default for all of our long running tasks should be non-blocking IMO.

> default to wait_for_completion=false so by default huge payloads are not returned to the user

This would be the better default even for curl users in a 5.0 world where the above returns a task id result where they can track the progress.

I don't like the implied contextual defaults proposed in https://github.com/elastic/elasticsearch/pull/20461 do we as clients now need to forcefully send `?max_reported_bulk_failures=2147483647` or try to sync that parameter with `batch_size` as long as `max_reported_bulk_failures` is not explicitly set? 

If someone wants to play they can use `wait_for_completion=true` and `size=<small_number>` no need for another knob.

**Comment by nik9000** [comment]
> I don't like the implied contextual defaults proposed in #20461 do we as clients now need to forcefully send ?max_reported_bulk_failures=2147483647 or try to sync that parameter with batch_size as long as max_reported_bulk_failures is not explicitly set?

You could set it to `-1` and not worry about `batch_size`, but this is lame either way.

I'm really not sure what the right thing is here because I expect plenty of people to run reindex and friends with curl/sense and plenty also to run it in clients. I do think having a different nob is useful here because the curl/sense folks want the process to run quickly as well.

Maybe a better compromise is to default `max_reported_bulk_failures` to `MAX_INT` unless `human` is on the url, in which case switch it to `50` or something.

**Comment by Mpdreamz** [comment]
I think the assumption we are doing folks using curl sense a service defaulting to wait for completion is not correct. 

Fire a reindex in sense and it'll be hard to tell if a request is pending or not. We can add spinners but after sometime people will assume spin is indefinite and might fire another request. 

I'd much much rather get an instant response on the command line from curl with a task id can check on status repeatedly then firing a request which blocks my stdin for an X amount of time.

**Comment by clintongormley** [comment]
To repeat the comment I've just added on the PR https://github.com/elastic/elasticsearch/pull/20461#issuecomment-247341263

> I don't think we need an override setting here. Reindex aborts when it gets error messages, and there is no indication of how many more exceptions would be thrown if reindex had just continued. I can't imagine a use case where a user will programmatically process each error case.

> But the default for all of our long running tasks should be non-blocking IMO.

Now that we have task management, I agree.  However, this would be a breaking change.  I think we should probably use deprecation logging in 5.x to warn that this default will change in 6.0 and to specify `wait_for_completion` explicitly to silence the warnings.

**Comment by clintongormley** [comment]
We should also be returning task IDs for requests like `_force_merge` which are only backgrounded today, but with no way of tracking progress.

**Comment by nik9000** [comment]
> I think the assumption we are doing folks using curl or sense a service defaulting wait_for_completion=true is not correct.

I think `wait_for_completion` is orthogonal from this issue though? If a large number of failures end up in the index you still have to get them out somehow.

**Comment by Mpdreamz** [comment]
@clintongormley some of these API's are expiremental in 2.x i think its ok to break in 5.x rather then 6.x?

@nik9000 it is orthogonal sorry for dragging it in this discussion should have opened a new one, just saw them as part of the too many toggles discussion.

Just wondering now if this is not a more global `get task api` problem, can we introduce a `defailed=true` or `verbosity=level` on the get task api and each task's content writer will know what to do? This parameter could then be used in conjuction with `wait_for_completion=true` on long running API's.

**Comment by nik9000** [comment]
> @nik9000 it is orthogonal sorry for dragging it in this discussion should have opened a new one, just saw them as part of the too many toggles discussion.

Its cool! I'm just trying to separate issues as much as possible so I can get things done. But maybe that is dangerous because it means we're not thinking holistically about the API.


For what it is worth the tasks API *does* have a detailed setting but it is for fetching the running status. When storing task results we always store the details and don't have a way of not fetching them.
